File: day1/solution.py
import sys

def rotate(position, distance):
    rotations = 0
    while distance != 0:
        if distance < 0:
            position -= 1
            distance += 1
            if position < 0:
                position += 100
        elif distance > 0:
            position += 1
            distance -= 1
            if position > 99:
                position -= 100
        if position == 0:
            rotations += 1
    return position, rotations


def part1():
    with open('input.txt') as f:
        zeroes = 0
        position = 50
        for instruction in f:
            distance = int(instruction[1:])
            if instruction.startswith('L'):
                distance = -distance
            position, _ = rotate(position, distance)
            if position == 0:
                zeroes += 1
        return zeroes

def part2():
    with open('input.txt') as f:
        zeroes = 0
        position = 50
        for instruction in f:
            distance = int(instruction[1:])
            if instruction.startswith('L'):
                distance = -distance
            position, rotations = rotate(position, distance)
            zeroes += rotations
        return zeroes

if __name__ == '__main__':
    if len(sys.argv) == 2 and sys.argv[1] == '2':
        print(part2())
    else:
        print(part1())

File: day1/solution.rb
# frozen_string_literal: true

def rotate(position, distance)
  rotations = 0
  until distance.zero?
    if distance.negative?
      position -= 1
      distance += 1
      position += 100 if position.negative?
    elsif distance.positive?
      position += 1
      distance -= 1
      position -= 100 if position > 99
    end
    rotations += 1 if position.zero?
  end
  [position, rotations]
end

def part1
  zeroes = 0
  position = 50
  File.open('input.txt').readlines.each do |line|
    distance = line.strip.chars[1..].join.to_i
    distance = -distance if line.start_with?('L')
    position, = rotate(position, distance)
    zeroes += 1 if position.zero?
  end
  zeroes
end

def part2
  zeroes = 0
  position = 50
  File.open('input.txt').readlines.each do |line|
    distance = line.strip.chars[1..].join.to_i
    distance = -distance if line.start_with?('L')
    position, rotations = rotate(position, distance)
    zeroes += rotations
  end
  zeroes
end

ARGV[0] == '2' ? puts(part2) : puts(part1)
